subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java

```diff
@@ -102,11 +105,68 @@
  */
 @DisableCachingByDefault(because = "Abstract super-class, not to be instantiated directly")
 public abstract class AbstractTestTask extends ConventionTask implements VerificationTask, Reporting<TestTaskReports> {
+
+    /**
+     * Wraps a list of listeners to subscribe, and lazily configures an anonymous broadcaster with those listeners when requested.
+     * Instances of this class are suitable for serialization (as long as listeners are serializable as well).
+     */
+    private class BroadcastSubscriptions<T> {
```

Comment:
🤔 This does solve the problem while preserving behavior but I'm wondering if we couldn't make the state even simpler... For example, does the task really need to support mutable subscriptions at execution time?

Reply:
I didn't have that at first, but seemed like such a cheap addition (maintenance-wise), that I thought would be worthwhile for the sake of allowing execution-time listeners to be registered.

Reply:
I can easily go back and remove support for adding listeners at execution time, and that would address my concerns with thread safety as well. I would feel inclined also to add a hard check preventing listeners from being added/removed by clients at execution time (instead of silently failing to notify them).

Reply:
> I would feel inclined also to add a hard check preventing listeners from being added/removed by clients at execution time (instead of silently failing to notify them).

Absolutely 👍 

That's a potentially breaking change though so we should better understand the impact. I'd start by adding the assertion that (external) listeners shouldn't be added at execution time and see what the smoke tests say.